aecore_eqc/txs_eqc.erl

```diff
@@ -2319,18 +2325,23 @@ fake_contract_id() ->
                   }).
 
 
-setup_data_dir(Prop) ->
+setup_data_dir() ->
     %% make sure we can run in eqc/aecore_eqc
-    ?SETUP(fun() ->
-                   {ok, Dir} = file:get_cwd(),
-                   DataDir = application:get_env(setup, data_dir),
-                   case lists:reverse(filename:split(Dir)) of
-                       [_, "eqc" | _] ->
-                           application:set_env(setup, data_dir, "../../data");
-                       _ ->
-                           application:set_env(setup, data_dir, "data")
-                   end,
-                   fun() ->
-                           application:set_env(setup, data_dir, DataDir)
-                   end
-           end, Prop).
+    {ok, Dir} = file:get_cwd(),
+    undefined = application:get_env(setup, data_dir),
+    case lists:reverse(filename:split(Dir)) of
+        [_, "eqc" | _] ->
+            application:set_env(setup, data_dir, "../../data");
+        _ ->
+            application:set_env(setup, data_dir, "data")
+    end,
+    fun() ->
+            ok = application:unset_env(setup, data_dir)
+    end.
+
+setup_hard_forks(X) ->
```

Comment:
This is not how SETUP in QuickCheck works. I believe you could nest SETUP, but you cannot get the teardown function as a return of a function in the setup sequence. Teardown is the thing the complete setup must return.

So you could probably write a setup and a teardown for both and put those in SETUP. Alternatively try to nest SETUP.

Reply:
> ... So you could probably write a setup and a teardown for both and put those in SETUP. ...

I believe this is how the PR is already doing. Could you please double check?

```erlang
    fun() ->
            _ = application:load(aecore),
            HardForksTeardown = setup_hard_forks(#{<<"1">> => 0, <<"2">> => Fork, <<"3">> => 2*Fork}),
            DataDirTeardown = setup_data_dir(),
            fun() ->
                    DataDirTeardown(),
                    HardForksTeardown()
            end
    end,
```